### ISSUE TITLE
[BUG] Internal Server Error on loading suggestions

### ISSUE THREAD

**Issue by steffen-heublein** [paperless-ngx/paperless-ngx#3441]
### Description

When opening a document the browser freezes for several seconds. On some occasions I get the following error message

```paperless Error retrieving suggestions: {"headers":{"normalizedNames":{},"lazyUpdate":null},"status":500,"statusText":"Internal Server Error","url":"http://192.168.1.102:8000/api/documents/1422/suggestions/","ok":false,"name":"HttpErrorResponse","message":"Http failure response for http://192.168.1.102:8000/api/documents/1422/suggestions/: 500 Internal Server Error","error":"\n<!doctype html>\n<html lang=\"en\">\n<head>\n <title>Server Error (500)</title>\n</head>\n<body>\n <h1>Server Error (500)</h1><p></p>\n</body>\n</html>\n"}```

This is also reflected when looking at the browser console which shows a response time for `http://192.168.1.102:8000/api/documents/1422/suggestions/` of about 12 seconds. The browser is unresponsive until this completes.

Paperless-ngx is run on an Unraid system inside their docker templating system. I understand that this is not the typical use case but maybe the error is enough to point me into the right direction. 

If I could switch off the suggestions API I guess the response times would also improve - this would be ok for me as a workaround.

### Steps to reproduce

1. Open any document

### Webserver logs

The logs from Paperless-ngx show nothing suspicious.

```bash
[2023-05-22 11:29:17,661] [DEBUG] [paperless.matching] Correspondent Deutsche Rentenversicherung matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains all of these words: Deutsche Rentenversicherung

[2023-05-22 11:29:17,666] [DEBUG] [paperless.matching] Correspondent Company matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains this word: Company

[2023-05-22 11:29:17,680] [DEBUG] [paperless.matching] Tag Person matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains all of these words: Person

[2023-05-22 11:29:17,682] [DEBUG] [paperless.matching] Tag Versicherung matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains this word: Versicherung

[2023-05-22 11:29:17,736] [DEBUG] [paperless.matching] Correspondent Deutsche Rentenversicherung matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains all of these words: Deutsche Rentenversicherung

[2023-05-22 11:29:17,741] [DEBUG] [paperless.matching] Correspondent Company matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains this word: Company

[2023-05-22 11:29:17,755] [DEBUG] [paperless.matching] Tag Person matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains all of these words: Person

[2023-05-22 11:29:17,756] [DEBUG] [paperless.matching] Tag Versicherung matched on document 2022-02-01 Deutsche Rentenversicherung Brother_ADS_1700W_20190213_201708_001470 because it contains this word: Versicherung```
```

(some words changed for privacy reasons)

### Paperless-ngx version

1.14.5

### Host OS

Unraid

### Installation method

Other (please describe above)

**Comment by stumpylog** [comment]
You can enable debug mode, but I doubt it would shed more light.  An internal error without any log seems more like a Django problem.  They'd be the ones sending the response in this case.

You could try to re-build the classification, perhaps the saved file has become corrupted.

**Comment by shamoon** [comment]
I suspect the frontend is now surfacing errors that may or may not be of actual consequence. If you look at the network call to `.../suggestions` in your browser (or just load it directly) what is the actual response?

https://github.com/paperless-ngx/paperless-ngx/discussions/3372
https://github.com/paperless-ngx/paperless-ngx/discussions/3428

**Comment by piraterx** [comment]
I started having same issue as well. When you try to edit a file to make changes, that's when happens mostly from sorting queue. But again, it is random.

**Comment by steffen-heublein** [comment]
@shamoon 
The endpoint response is looking fine to me:

Header:
```
HTTP/1.1 200 OK
date: Tue, 23 May 2023 12:39:32 GMT
server: uvicorn
content-type: application/json
vary: Accept, Accept-Language, Origin, Cookie
allow: GET, HEAD, OPTIONS
x-frame-options: SAMEORIGIN
x-api-version: 3
x-version: 1.14.5
content-length: 119
content-language: de-de
x-content-type-options: nosniff
referrer-policy: same-origin
cross-origin-opener-policy: same-origin
```

Content:
```
{
	"correspondents":[1,41],
	"tags":[22,28],
	"document_types":[24,1],
	"storage_paths":[],
	"dates":["2023-04-17","2023-04-24"]
}
```

The response just takes a really long time. I'm wondering why it's blocking the browser. Can the suggestions feature be switched off somehow?

The issues that you linked look similar to what I am seeing.

**Comment by shamoon** [comment]
The speed will depend on a few things including your system and the pdf itself, searching the document for dates in particular can be a bit taxing.

You can try lowering [`PAPERLESS_NUMBER_OF_SUGGESTED_DATES`](https://docs.paperless-ngx.com/configuration/) to see if that helps.

Again, overall Im not sure how actionable this is at the moment. I've never been able to reproduce this error message, it does seem to be intermittent but also kind of inconsequential in the end. The next release will make it easier to see the full output of the error and if we can get a bit more detail perhaps we can narrow it down.

**Comment by polgarc** [comment]
I've seen the same for a few times. It happens when I scan a document, I'm on the dashboard, waiting for the processing to be finished. Once it's done (progress 100%), I click on the link to go to the new document's page. The page loads as usual, but there's this red warning with error `500`.  
I've tried to reproduce it, but not always happens. 

OP mentioned long response time - in my case it loads fast.  
Also I use Docker + Firefox.

Stacktrace:

```
[2023-05-23 20:22:57,825] [ERROR] [django.request] Internal Server Error: /api/documents/407/suggestions/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 486, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 43, in inner
    response = await get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
    response = await wrapped_callback(
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 448, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/usr/local/lib/python3.9/site-packages/asgiref/current_thread_executor.py", line 22, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 490, in thread_handler
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 55, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/src/paperless/src/documents/views.py", line 433, in suggestions
    c.id for c in match_correspondents(doc, classifier, request.user)
  File "/usr/src/paperless/src/documents/matching.py", line 23, in match_correspondents
    pred_id = classifier.predict_correspondent(document.content) if classifier else None
  File "/usr/src/paperless/src/documents/classifier.py", line 363, in predict_correspondent
    X = self.data_vectorizer.transform([self.preprocess_content(content)])
  File "/usr/src/paperless/src/documents/classifier.py", line 348, in preprocess_content
    self._stop_words = set(stopwords.words(settings.NLTK_LANGUAGE))
  File "/usr/local/lib/python3.9/site-packages/nltk/corpus/util.py", line 121, in __getattr__
    self.__load()
  File "/usr/local/lib/python3.9/site-packages/nltk/corpus/util.py", line 95, in __load
    args, kwargs = self.__args, self.__kwargs
AttributeError: 'WordListCorpusReader' object has no attribute '_LazyCorpusLoader__args'
```

**Comment by shamoon** [comment]
Thanks for the stack trace. That’s the same as one of the [other threads](https://github.com/paperless-ngx/paperless-ngx/issues/3392) though in looking into it I remember it was a sort of inexplicable error. It might be worth wrapping the nltk logic with a `try` and just logging an error rather than flub the whole response

**Comment by adamsocrat** [comment]
```
[2023-05-29 20:12:49,057] [ERROR] [django.request] Internal Server Error: /api/documents/15/suggestions/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 486, in thread_handler
    raise exc_info[1]
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 43, in inner
    response = await get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
    response = await wrapped_callback(
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 448, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/usr/local/lib/python3.9/site-packages/asgiref/current_thread_executor.py", line 22, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 490, in thread_handler
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 55, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/src/paperless/src/documents/views.py", line 435, in suggestions
    "tags": [t.id for t in match_tags(doc, classifier, request.user)],
  File "/usr/src/paperless/src/documents/matching.py", line 63, in match_tags
    predicted_tag_ids = classifier.predict_tags(document.content) if classifier else []
  File "/usr/src/paperless/src/documents/classifier.py", line 387, in predict_tags
    X = self.data_vectorizer.transform([self.preprocess_content(content)])
  File "/usr/src/paperless/src/documents/classifier.py", line 348, in preprocess_content
    self._stop_words = set(stopwords.words(settings.NLTK_LANGUAGE))
  File "/usr/local/lib/python3.9/site-packages/nltk/corpus/util.py", line 121, in __getattr__
    self.__load()
  File "/usr/local/lib/python3.9/site-packages/nltk/corpus/util.py", line 95, in __load
    args, kwargs = self.__args, self.__kwargs
AttributeError: 'WordListCorpusReader' object has no attribute '_LazyCorpusLoader__args' 
```

Had the same error. FYI I changed the document-type the moment it started consuming and then the error occurred. IDK if there is any relevance to that.

**Comment by mutax** [comment]
I get the same issue from time to time, when searching for the error I found something on stackoverflow.

https://stackoverflow.com/questions/27433370/what-would-cause-wordnetcorpusreader-to-have-no-attribute-lazycorpusloader

This might be a race condition between api calls / webpage loading in the browser. 

Shortly after the log states 
`[2023-06-08 18:07:45,304] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] lots of diacritics - possibly poor OCR`

Maybe these are related because preparing the object takes longer than usual and the request for the suggestions hits before it is complete.

**Comment by shamoon** [comment]
There was already a fix merged that will hopefully prevent this, though it's hard to say for certain since it cant always be reproduced. That change was included in > v1.15.0

https://github.com/paperless-ngx/paperless-ngx/pull/3453
